### PR TITLE
[Spark-8244][SQL] string function: find in set

### PR DESCRIPTION
**Pull request by tarekbecker**
Jira https://issues.apache.org/jira/browse/SPARK-8244

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 1]
```diff
@@ -211,6 +211,77 @@ case class EndsWith(left: Expression, right: Expression)
 }
 
 /**
+ * A function that returns the index (1-based) of the given string (left) in the comma-
+ * delimited list (right). Returns -1, if the string wasn't found and 0 if the given
+ * string (left) contains a comma.
+ */
+case class FindInSet(left: Expression, right: Expression) extends BinaryExpression
+    with ExpectsInputTypes {
```
Comment (tarekauel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Should this better implement `AutoCastInputTypes`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 2]
```diff
@@ -211,6 +211,77 @@ case class EndsWith(left: Expression, right: Expression)
...
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = left.eval(input)
+    if (valueLeft != null) {
+      val strLeft = valueLeft.asInstanceOf[UTF8String]
+      if (!strLeft.contains(UTF8String.fromString(","))) {
+        val valueRight = right.eval(input)
+        if (valueRight != null) {
+          val splits = valueRight.asInstanceOf[UTF8String].toString.split(",")
+          var i = 0
```
Comment (HuJiayin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: I think the code has performance issue. When you call split, it scans the string for one time. After that, you process the string for the second time. If the set is large, the performance issue will appear. You could scan the set by one time.

Comment (tarekauel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: That's a good comment. My feeling is that we should implement a `split` or `splitByString` method as part of `UTF8String`. (Could be a subtask of https://issues.apache.org/jira/browse/SPARK-8829). @rxin what do you think?

Comment (HuJiayin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: You don't need to implement a split in this issue. That is not the point. The point is you don't need to split.

Comment (tarekauel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Ah yes thanks. So I wait until #7197 got merged, that will make things easier.

Comment (HuJiayin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: I think #7197 doesn't relate your codes' low performance issue. You need to rewrite your implementation. Scan the set for one pass can solve this issue however you need modify your argorithms. The extra split step has both memory comsuption and low performance issue.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -206,6 +206,30 @@ public UTF8String toLowerCase() {
...
+   * -1 will be returned, else the index of match (1-based index)
+   */
+  public int findInSet(UTF8String match) {
+    if (match.contains(UTF8String.fromString(","))) {
+      return 0;
+    }
+
+    int lastComma = -1;
+    for (int i = 0, n = 1; i < numBytes; i++) {
+      if (getByte(i) == 44) {
```
Comment (tarekauel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: One comment to this. 44 as bitstring is `00101100` (http://www.fileformat.info/info/unicode/char/2c/index.htm). Just ASCII characters can start with `00`, see https://en.wikipedia.org/wiki/UTF-8.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 4]
```diff
@@ -207,6 +207,30 @@ public UTF8String toLowerCase() {
   }
 
   /**
+   * Returns the index of the string `match` in this String. This string has to be a comma separated
+   * list. If `match` contains a comma 0 will be returned. If the `match` isn't part of this String,
+   * -1 will be returned, else the index of match (1-based index)
+   */
+  public int findInSet(UTF8String match) {
+    if (match.contains(UTF8String.fromString(","))) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Could you create a static constant for `,` ?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 3 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Use `','.toByte` instead

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 5]
```diff
@@ -207,6 +207,30 @@ public UTF8String toLowerCase() {
...
+      return 0;
+    }
+
+    int lastComma = -1;
+    for (int i = 0, n = 1; i < numBytes; i++) {
+      if (getByte(i) == 44) {
+        if (i - (lastComma + 1) == match.numBytes && ByteArrayMethods.arrayEquals(base,
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: break at `&&`?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 6]
```diff
@@ -207,6 +207,30 @@ public UTF8String toLowerCase() {
...
+            offset + (lastComma + 1), match.base, match.offset, i - (lastComma + 1))) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: match.numBytes

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 7]
```diff
@@ -207,6 +207,30 @@ public UTF8String toLowerCase() {
...
+   * list. If `match` contains a comma 0 will be returned. If the `match` isn't part of this String,
+   * -1 will be returned, else the index of match (1-based index)
+   */
+  public int findInSet(UTF8String match) {
+    if (match.contains(UTF8String.fromString(","))) {
+      return 0;
+    }
+
+    int lastComma = -1;
+    for (int i = 0, n = 1; i < numBytes; i++) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: nit: move `n = 1` before the loop

File: unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 8]
```diff
@@ -248,4 +248,13 @@ public void levenshteinDistance() {
     assertEquals(
         UTF8String.fromString("世界千世").levenshteinDistance(UTF8String.fromString("千a世b")),4);
   }
+
+  @Test
+  public void findInSet() {
+    assertEquals(fromString("abc,b,ab,c,def").findInSet(fromString("ab")), 3);
+    assertEquals(fromString("ab,abc,b,ab,c,def").findInSet(fromString("ab")), 1);
+    assertEquals(fromString(",,,ab,abc,b,ab,c,def").findInSet(fromString("ab")), 4);
+    assertEquals(fromString(",ab,abc,b,ab,c,def").findInSet(fromString("")), 1);
```
Comment (davies) [inline unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java]: Could you add a test case for `"ab".find("ab")`

**Comment by tarekauel** [comment]
Jenkins, test this please.

**Comment by davies** [comment]
LGTM, could you update the description of PR?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 9]
```diff
@@ -50,6 +50,8 @@
     5, 5, 5, 5,
     6, 6, 6, 6};
 
+  private static UTF8String COMMA_UTF8 = UTF8String.fromString(",");
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Add `final`.

**Comment by chenghao-intel** [comment]
LGTM.

**Comment by tarekauel** [comment]
Jenkins, test this please.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 10]
```diff
@@ -231,6 +233,36 @@ public UTF8String toLowerCase() {
...
+    if (match.contains(COMMA_UTF8)) {
+      return 0;
+    }
+
+    int n = 1, lastComma = -1;
+    for (int i = 0; i < numBytes; i++) {
+      if (getByte(i) == (byte) ',') {
+        if (i - (lastComma + 1) == match.numBytes &&
+            ByteArrayMethods.arrayEquals(base, offset + (lastComma + 1), match.base, match.offset,
+                match.numBytes)) {
```
Comment (HuJiayin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: When you call the arrayEquals, the arrayEquals start from the position in the set your code already scanned, the overall time consumption will be 2 times than scan one time to get the result. This implement is similar with 1. split 2. check match. If the set the big, the time cost will be large. I think you can compare the byte onebyone and find the index in one execution.

**Comment by davies** [comment]
test this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 11]
```diff
@@ -516,4 +516,14 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
   }
+
+  test("find in set") {
+    checkEvaluation(
+      FindInSet(Literal.create(null, StringType), Literal.create(null, StringType)), null)
+    checkEvaluation(FindInSet(Literal("ab"), Literal.create(null, StringType)), null)
+    checkEvaluation(FindInSet(Literal.create(null, StringType), Literal("abc,b,ab,c,def")), null)
+    checkEvaluation(FindInSet(Literal("ab"), Literal("abc,b,ab,c,def")), 3)
+    checkEvaluation(FindInSet(Literal("abf"), Literal("abc,b,ab,c,def")), 0)
+    checkEvaluation(FindInSet(Literal("ab,"), Literal("abc,b,ab,c,def")), 0)
```
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: Could we have a special case for empty string?

Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: I see it in UTF8StringSuite, make sense.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 10 above)
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: The current approach have the optimization of check the length of match first, so it actually will not call arrayEquals many times in practice. It's good to me.

**Comment by davies** [comment]
LGTM, @tarekauel Could you fix the conflict?

**Pull request closed without merging**